Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp

```diff
@@ -187,9 +204,59 @@ void ActsExamples::SensitiveSteppingAction::UserSteppingAction(
     return;
   }
 
+  // Output is only strictly valid if step logging is not enabled
   const auto particleId = eventStore().trackIdMapping.at(track->GetTrackID());
-
-  ACTS_VERBOSE("Step of " << particleId << " in sensitive volume " << geoId);
+  if (!m_cfg.stepLogging && surface != nullptr) {
+    ACTS_VERBOSE("Step of " << particleId << " in sensitive volume " << geoId);
+  } else if (m_cfg.stepLogging) {
+    if (!eventStore().propagationRecords.contains(track->GetTrackID())) {
+      // Create the propagation summary
+      Acts::ActsScalar xVtx = track->GetVertexPosition().x();
+      Acts::ActsScalar yVtx = track->GetVertexPosition().y();
+      Acts::ActsScalar zVtx = track->GetVertexPosition().z();
+      Acts::ActsScalar xDirVtx = track->GetVertexMomentumDirection().x();
+      Acts::ActsScalar yDirVtx = track->GetVertexMomentumDirection().y();
+      Acts::ActsScalar zDirVtx = track->GetVertexMomentumDirection().z();
+      Acts::ActsScalar absMomentum = track->GetMomentum().mag();
+
+      PropagationSummary iSummary(Acts::CurvilinearTrackParameters(
+          Acts::Vector4(xVtx, yVtx, zVtx, 0.),
+          Acts::Vector3(xDirVtx, yDirVtx, zDirVtx), absCharge / absMomentum,
+          std::nullopt, Acts::ParticleHypothesis::pion()));
+
+      eventStore().propagationRecords.insert({track->GetTrackID(), iSummary});
+    }
+    PropagationSummary& pSummary =
+        eventStore().propagationRecords.at(track->GetTrackID());
+
+    // Increase the step counter
+    pSummary.nSteps += 1;
+
+    Acts::ActsScalar currentTrackLength =
+        track->GetTrackLength() * convertLength;
+    Acts::ActsScalar currentStepLength =
+        currentTrackLength - pSummary.pathLength;
+    pSummary.pathLength = currentTrackLength;
+
+    // Create a new step for the step logging
+    Acts::detail::Step pStep = stepFromG4Step(step);
+    pStep.geoID = geoId;
+    pStep.surface = surface != nullptr ? surface->getSharedPtr() : nullptr;
+    // Check if last step was on same surface
+    if (!pSummary.steps.empty() && pSummary.steps.back().geoID == geoId &&
+        pSummary.steps.back().surface != nullptr) {
+      auto& lastStep = pSummary.steps.back();
+      lastStep.stepSize = Acts::ConstrainedStep(currentStepLength);
+      lastStep.position = 0.5 * (pStep.position + lastStep.position);
```

Comment:
same here

Reply:
See my comment, I don't actually know what exactly we should do if we wanted to also get the range kutta steps, I will investigate

Reply:
I am checking with John Apostolakis from Geant4 so we understand exactly what a Geant4 step is and how we can deduce our step definition from it.